### PR TITLE
make bias configurable

### PR DESCRIPTION
**Pull request by ngoyal2707**
made biases configurable.
Most latest models dont train with biases (palm, llama).

Although ideally I'd like do disable layernorm bias also with the same config param

### CODE REVIEW

**Comment by ngoyal2707** [comment]
cc: @ptrendx @ksivaman appreciate if you can take a look, all tests pass

File: transformer_engine/pytorch/module.py [hunk 1]
```diff
@@ -2147,7 +2147,7 @@ def forward(
                 activation_dtype,
                 get_workspace(),
                 bias=fc1_bias,
-                use_bias=True,
```
Comment (timmoon10) [inline transformer_engine/pytorch/module.py:2150]: As far as I can tell, this old behavior of always setting `bias=True` for FC1 in `LayerNormMLP` was a discrepancy from the Megatron reference. This PR seems to match the reference.

**Comment by ngoyal2707** [comment]
@timmoon10 thanks added your suggested changes

File: transformer_engine/pytorch/jit.py [hunk 2]
```diff
@@ -55,18 +63,39 @@ def bgrad_dgelu_fused_(
     return bgrad, dgelu
 
 
+@torch.jit.script
+def dgelu_fused_(
+    grad_output: torch.Tensor, inp: torch.Tensor
+) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (ksivaman) [inline transformer_engine/pytorch/jit.py:69]:
```suggestion
) -> torch.Tensor:
```

Comment (ksivaman) [reply transformer_engine/pytorch/jit.py:69]: Jit compilation fails w wrong annotation, this is the fix

File: transformer_engine/pytorch/jit.py [hunk 3]
```diff
@@ -35,6 +35,14 @@ def bias_gelu_fused_(inp: torch.Tensor, bias: torch.Tensor) -> torch.Tensor:
     x = inp + bias
     return x * 0.5 * (1.0 + torch.tanh(0.79788456 * x * (1 + 0.044715 * x * x)))
 
+@torch.jit.script
```
Comment (ksivaman) [inline transformer_engine/pytorch/jit.py:38]: This needs to be wrapped around `torch.cuda.amp.autocast(enabled=False)` similar to the other fused functions in order to work with AMP. Otherwise the generated kernel will return an fp32 output regardless.

File: transformer_engine/pytorch/jit.py [hunk 4]
```diff
@@ -55,18 +63,39 @@ def bgrad_dgelu_fused_(
     return bgrad, dgelu
 
 
+@torch.jit.script
```
Comment (ksivaman) [inline transformer_engine/pytorch/jit.py:66]: Same comment as above

File: transformer_engine/pytorch/jit.py (hunk 3 above)
Comment (ngoyal2707) [reply transformer_engine/pytorch/jit.py:38]: both are already wrapped in `torch.cuda.amp.autocast(enabled=False)` in `bias_gelu_fused` and `bgrad_dgelu_fused`

**Comment by ngoyal2707** [comment]
@ksivaman sorry didnt get your comment, both are already wrapped in the wrapper function.

File: transformer_engine/pytorch/jit.py (hunk 3 above)
Comment (ksivaman) [reply transformer_engine/pytorch/jit.py:38]: I missed that, thanks

File: transformer_engine/pytorch/jit.py [hunk 5]
```diff
@@ -55,18 +63,39 @@ def bgrad_dgelu_fused_(
...
 
 def bgrad_dgelu_fused(
     grad_output: torch.Tensor, inp: torch.Tensor, bias: torch.Tensor
 ) -> Tuple[torch.Tensor, torch.Tensor]:
     """Disable native AMP for `bgrad_dgelu_fused_`"""
     with torch.cuda.amp.autocast(enabled=False):
-        return bgrad_dgelu_fused_(grad_output, inp, bias)
+        if bias.numel() != 0:
+            return bgrad_dgelu_fused_(grad_output, inp, bias)
+        return torch.zeros_like(bias), dgelu_fused_(grad_output, inp)
```
Comment (ksivaman) [inline transformer_engine/pytorch/jit.py:98]: Is this additional allocation necessary? Why don't we return `None` here and change the return type to `Tuple[Union[torch.Tensor, None], torch.Tensor]`?

File: transformer_engine/pytorch/transformer.py [hunk 6]
```diff
@@ -494,6 +494,7 @@ def __init__(
         fuse_qkv_params: bool = False,
         zero_centered_gamma: bool = False,
         qkv_weight_interleaved: bool = True,
+        bias: bool = True,
```
Comment (ksivaman) [inline transformer_engine/pytorch/transformer.py:497]: This would need documentation

File: transformer_engine/pytorch/module.py [hunk 7]
```diff
@@ -2175,7 +2175,7 @@ def forward(
...
+                use_bias=use_bias,
```
Comment (ksivaman) [inline transformer_engine/pytorch/module.py:2178]:
```suggestion
                use_bias=fc1_bias.numel() > 0,
```

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:2178]: `use_bias` cannot be used here since it is meant for `fc2_bias`.

File: transformer_engine/pytorch/module.py [hunk 8]
```diff
@@ -2224,12 +2226,13 @@ def forward(
                 activation_dtype,
                 get_workspace(),
                 bias=fc1_bias,
-                use_bias=not bias_gelu_nvfusion,
+                use_bias=(not bias_gelu_nvfusion) and use_bias,
```
Comment (ksivaman) [inline transformer_engine/pytorch/module.py:2229]:
```suggestion
                use_bias=(not bias_gelu_nvfusion) and fc1_bias.numel() > 0,
```

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:2229]: Same reason as above comments. However, this is a bit confusing. Switching fully to `use_fc1_bias` and `use_fc2_bias` is better.

File: transformer_engine/pytorch/jit.py (hunk 5 above)
Comment (timmoon10) [reply transformer_engine/pytorch/jit.py:98]: It seems logical to me that `bias` and `bgrad` should have the same type, although I'm not too committed to it. I think the "right" answer would be for `bias` to be `Optional[torch.Tensor]`, but that's probably beyond the scope of this PR.

Comment (ksivaman) [reply transformer_engine/pytorch/jit.py:98]: Re: `bias` and `bgrad` should have the same type
I agree with this when a gradient is being computed but that's not the case here right.


Re: the "right" answer would be for `bias` to be `Optional[torch.Tensor]`
While I agree with this it would require modifications at the call site to reflect this.

**Review by timmoon10 — APPROVED** [review]
LGTM

**Review by ksivaman — APPROVED** [review]
Thanks @ngoyal2707!

File: transformer_engine/common/rmsnorm/rmsnorm_api.cpp [hunk 9]
```diff
@@ -97,7 +97,7 @@ BwdFunction &get_bwd_launcher(DType wtype, DType itype, DType otype, DType ctype
 // ////////////////////////////////////////////////////////////////////////////////////////////////////
 
 inline size_t product(const std::vector<size_t> &shape) {
-    return std::reduce(shape.cbegin(), shape.cend(), size_t{1}, std::multiplies<>());
+    return std::accumulate(shape.cbegin(), shape.cend(), size_t{1}, std::multiplies<>());
```
Comment (awgu) [inline transformer_engine/common/rmsnorm/rmsnorm_api.cpp:100]: Curious: Why use `std::accumulate` instead of `std::reduce`?
